File: services/identity-service/src/main/java/br/com/deltaglobalbank/identity/infrastructure/security/apikey/ApiKeyGenerator.java
package br.com.deltaglobalbank.identity.infrastructure.security.apikey;

import java.security.SecureRandom;

import org.springframework.stereotype.Component;

@Component
public class ApiKeyGenerator {

    public static final int KEY_RANDOM_LENGTH = 32;
    public static final int PREFIX_LENGTH = 13;

    private static final String ALPHABET = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

    private final ApiKeyProperties properties;
    private final SecureRandom secureRandom = new SecureRandom();

    public ApiKeyGenerator(ApiKeyProperties properties) {
        this.properties = properties;
    }

    public GeneratedApiKey generate() {
        StringBuilder random = new StringBuilder(KEY_RANDOM_LENGTH);
        for (int i = 0; i < KEY_RANDOM_LENGTH; i++) {
            random.append(ALPHABET.charAt(secureRandom.nextInt(ALPHABET.length())));
        }
        String plainKey = properties.brand() + "_" + properties.environment() + "_" + random;
        String prefix = plainKey.substring(0, PREFIX_LENGTH);

        return new GeneratedApiKey(plainKey, prefix);
    }
}

File: services/identity-service/src/main/kotlin/br/com/deltaglobalbank/identity/infrastructure/security/apikey/ApiKeyGenerator.kt
package br.com.deltaglobalbank.identity.infrastructure.security.apikey

import org.springframework.stereotype.Component
import java.security.SecureRandom

@Component
class ApiKeyGenerator(private val properties: ApiKeyProperties) {
    private val secureRandom = SecureRandom()
    private val alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"

    fun generate(): GeneratedApiKey {
        val random = buildString(KEY_RANDOM_LENGTH) {
            repeat(KEY_RANDOM_LENGTH) {
                append(alphabet[secureRandom.nextInt(alphabet.length)])
            }
        }
        val plainKey = "${properties.brand}_${properties.environment}_$random"
        val prefix = plainKey.substring(0, PREFIX_LENGTH)

        return GeneratedApiKey(
            plainKey = plainKey,
            prefix = prefix,
        )
    }

    companion object {
        const val KEY_RANDOM_LENGTH = 32
        const val PREFIX_LENGTH = 13
    }
}

data class GeneratedApiKey(
    val plainKey: String,
    val prefix: String,
)